### ISSUE TITLE
add realloc_array function

### ISSUE THREAD

**Issue by goatshriek** [goatshriek/stumpless#443]
Stumpless includes wrappers for memory allocation to allow a customized memory allocator to be used. One of these is `realloc_mem`, which calls the underlying allocator's `realloc` routine.

As outlined in #440, there are situations where a multiplication is done to get the new size. For the same reasons outlined in that issue, it would also be helpful to add a `realloc` wrapper that does overflow checks.

## General Approach
There are a few details left out of the following approach, for you to fill in as you encounter them. If you find you need help, please ask here or on the [project gitter](https://gitter.im/stumpless/community) and someone can help you get past the stumbling block.

First, review #440 and understand the solution proposed there; this issue is very similar. **If you solved the `alloc_array` issue yourself, please leave this one for a new contributor. This is a "good first issue", and as such the learning experience is an essential component.**

Add a new function to `src/memory.c` named `realloc_array` with the following signature:

```c
void *
realloc_array( const void *mem, size_t item_count, size_t item_size );
```

Once you've made this addition, swap out the usage of `realloc_mem` with multiplication in `src/cache.c` `add_page` and `src/target/journald.c` `init_fields`, and make sure that all tests continue to pass.

Next, refactor  `stumpless_add_param` in `src/element.c` to use the new `realloc_array` function. You can then create a new test in `test/function/element.cpp` that modifies the `param_count` of the element structure to be sufficiently high (`SIZE_MAX` perhaps?) to cause an overflow and subsequent memory allocation failure to test your detection logic.

**Comment by Brijeshthummar02** [comment]
i can fix this.

**Comment by m3g4d1v3r** [comment]
Hello there!

I've created a pull request that aims to solve this issue: #472

**Comment by ansh-dulewale** [comment]
I can fix this.

**Comment by goatshriek** [comment]
Pull #472 already addresses this and is near completion; it would be better to find a different issue to work on.

**Comment by goatshriek** [comment]
Resolved in #472.
